sycl/source/detail/scheduler/scheduler.cpp

```diff
@@ -138,7 +138,10 @@ std::vector<EventImplPtr> Scheduler::getWaitList(EventImplPtr Event) {
 
 void Scheduler::waitForEvent(EventImplPtr Event) {
   std::shared_lock<std::shared_timed_mutex> Lock(MGraphLock);
-  GraphProcessor::waitForEvent(std::move(Event));
+  // The command might have been cleaned up (and set to nullptr) by another
+  // thread
+  if (Event->getCommand())
```

Comment:
I believe the pointer to command in `event_impl` should be atomic, then.

Reply:
Is this necessary, since command cleanup is mutex-controlled in all cases (2 currently: `Scheduler::cleanupFinishedCommands` and `Scheduler::removeMemoryObject`)?

Reply:
Even though the cleanup (write) is mutex-controlled, read of this command isn't always mutex controlled.

I believe [this one](https://github.com/intel/llvm/blob/433c14be9644855c688c59c9b03bcc44a447f020/sycl/source/detail/scheduler/graph_builder.cpp#L942) and [that](https://github.com/intel/llvm/blob/767ddfe9343ba38639d09b107c7afdc6f2a6dcfb/sycl/source/detail/scheduler/commands.cpp#L504) are executed with write lock. [Scheduler::GraphProcessor::getWaitList()](https://github.com/intel/llvm/blob/ae3fd5c48bb55176016a7c7c6fa2ac5684aaff55/sycl/source/detail/scheduler/graph_processor.cpp#L26) is executed with read lock. Not sure if [Scheduler::GraphProcessor::waitForEvent()](https://github.com/intel/llvm/blob/ae3fd5c48bb55176016a7c7c6fa2ac5684aaff55/sycl/source/detail/scheduler/graph_processor.cpp#L40) is executed with any lock on.

Reply:
Actually `Scheduler::GraphProcessor::waitForEvent()` is not executed with any lock on from withing [Scheduler::waitForRecordToFinish()](https://github.com/intel/llvm/blob/ebace7711416d761a36f3c5dfff9327d1bf81894/sycl/source/detail/scheduler/scheduler.cpp#L23).

Reply:
`Scheduler::waitForRecordToFinish` is only called from `Scheduler::removeMemoryObject`, with write lock. I will add comments to use Scheduler mutex to `event_impl::getCommand`/`event_impl::setCommand` descriptions.